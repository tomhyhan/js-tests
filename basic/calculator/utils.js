function add(x,y) {
    return x+y
}

function sub(x,y) {
    return x-y
}

function mul(x,y) {
    return x*y
}

function div(x,y) {
    if (y == 0) {
        throw new Error("Can not divide by 0")
    }
    return x / y
}

module.exports = {add, sub, div, mul}