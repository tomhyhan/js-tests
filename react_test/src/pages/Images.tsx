import React from 'react'
import useSWR from 'swr'
import { useImage } from '../context/ImageProvider'
import Image from '../component/Image'
import { ImageT } from '../lib/Types/ImageT'
import AddImage from '../component/AddImage'

export default function Images() {
    const {imageService} = useImage()
    const {data: images, error, isLoading , mutate} = useSWR("images", () => imageService.getImages(), {
        revalidateOnFocus: false,
    })

    const handleImageAdd = async (imageLayout:string) => {
        const newImage = await imageService.getImage(imageLayout)
        await mutate([...images, newImage], {
            revalidate:false
        })
    }

    return (
        <main>
            <AddImage onAddImage={handleImageAdd} />
            {isLoading && <div>Loading...</div>}
            {error && <div>Something Went Wrong!</div>}
            <ul className="grid grid-cols-2 gap-4">
                {images?.map((image: ImageT) => (
                    <Image key={image.id} image={image}/>
                ))}
            </ul>
        </main>
  )
}
