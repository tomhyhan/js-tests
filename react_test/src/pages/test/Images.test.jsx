import { render, screen, waitFor } from '@testing-library/react';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { withRouter, withContext } from '../../test/utils';
import { Route } from 'react-router-dom';
import Images from '../Images';
import { fakeImages } from '../../test/fakeData';
import '@testing-library/jest-dom';

// renders images
// showing loading
// showing error
// adding normal image
// adding blur image
// adding grayscale image

describe('Images', () => {
  const fakeImageService = {
    getImages: vi.fn(),
  };

  beforeEach(() => {
    fakeImageService.getImages.mockImplementation(() => fakeImages);
  });

  afterEach(() => {
    fakeImageService.getImages.mockReset();
  });

  it('renders all Images list', async () => {
    const { asFragment } = render(
      withContext(
        withRouter(<Route path='/' element={<Images />} />),
        fakeImageService
      )
    );

    expect(fakeImageService.getImages).toHaveBeenCalledOnce();

    await waitFor(() =>
      expect(screen.getAllByRole('listitem')).toHaveLength(fakeImages.length)
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders loading state', async () => {
    const { debug } = render(
      withContext(
        withRouter(<Route path='/' element={<Images />} />),
        fakeImageService
      )
    );
    console.log('debug');
    console.log(debug());
    expect(screen.getByText(/Loading.../)).toBeInTheDocument();
  });
});
