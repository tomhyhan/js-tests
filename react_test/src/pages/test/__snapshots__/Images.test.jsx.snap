// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Images > renders all Images list 1`] = `
<DocumentFragment>
  <main>
    <form
      class="flex p-5"
    >
      <select
        class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded block mr-5"
        name="imageLayout"
      >
        <option
          selected=""
          value="normal"
        >
          normal
        </option>
        <option
          value="blur"
        >
          blur
        </option>
        <option
          value="grayscale"
        >
          grayscale
        </option>
      </select>
      <button
        class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
        type="submit"
      >
        Add Image
      </button>
    </form>
    <ul
      class="grid grid-cols-2 gap-4"
    >
      <li
        class="px-4 cursor-pointer"
      >
        <img
          alt="Foo"
          src="https://foo.bar.download_url"
        />
      </li>
      <li
        class="px-4 cursor-pointer"
      >
        <img
          alt="john"
          src="https://john.doe.download_url"
        />
      </li>
    </ul>
  </main>
</DocumentFragment>
`;
